Add a mechanism to schedule opportunistic work in between rendering updates

whsieh opened the pull request:
#### c7bdb5bcd411cab772466fdc79661c41c115095a
<pre>
Add a mechanism to schedule opportunistic work in between rendering updates
<a href="https://bugs.webkit.org/show_bug.cgi?id=258032">https://bugs.webkit.org/show_bug.cgi?id=258032</a>
rdar://110076652

Reviewed by Ryosuke Niwa.

Add an `OpportunisticTaskScheduler` helper class to be used in a subsequent patch to dispatch
periodically-scheduled cleanup tasks, such as incremental sweeping, in between rendering updates
(i.e. after the runloop begins to wait for more work, but before the next rendering update is
scheduled to start).

This `OpportunisticTaskScheduler` is essentially a wrapper around a `RunLoopObserver`, which calls
into `Page` to perform scheduled work until a given deadline. The `Page` is the only entity that
owns the task scheduler, and will tell it to reschedule after every rendering update. Additionally,
we introduce another mechanism — `OpportunisticTaskDeferralScope` — that prevents the task scheduler
from firing during certain time intervals where we know that scheduled work is imminent. For now,
this just includes:

1. Scheduled one-shot timers with a very short duration.
2. Any time prior to first paint, after navigation.

No change in behavior (yet).

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ScriptExecutionContext.h:
(WebCore::ScriptExecutionContext::addTimeout):
(WebCore::ScriptExecutionContext::takeTimeout):
(WebCore::ScriptExecutionContext::removeTimeout): Deleted.

Change `removeTimeout` to `takeTimeout`, and have it return the removed `DOMTimer`. This allows us
to (more easily) call `clearOpportunisticTaskDeferralScopeIfPossible()` on the removed timer below.

* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimer::install):
(WebCore::DOMTimer::removeById):
(WebCore::DOMTimer::fired):
(WebCore::DOMTimer::didStop):
(WebCore::DOMTimer::makeOpportunisticTaskDeferralScopeIfPossible):
(WebCore::DOMTimer::clearOpportunisticTaskDeferralScopeIfPossible):

Add a opportunistic task deferral scope to prevent us from scheduling opportunistic tasks in the
case where one-shot zero-delay timers are about to fire.

* Source/WebCore/page/DOMTimer.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::fireLayoutRelatedMilestonesIfNeeded):

Add some plumbing to notify `Page` when we hit first meaningful paint.

* Source/WebCore/page/OpportunisticTaskScheduler.cpp: Added.
(WebCore::OpportunisticTaskDeferralScope::OpportunisticTaskDeferralScope):
(WebCore::OpportunisticTaskDeferralScope::~OpportunisticTaskDeferralScope):

Add a RAII helper class to increment and decrement the deferral count in
`OpportunisticTaskScheduler`, which prevents opportunistically scheduled tasks from firing; this
RAII pattern hopefully makes it a bit easier to ensure that nothing ends up with unbalanced calls to
increment or decrement.

(WebCore::OpportunisticTaskScheduler::OpportunisticTaskScheduler):
(WebCore::OpportunisticTaskScheduler::reschedule):

Helper method to reschedule the task scheduler, with a target deadline; called after completing
rendering updates in `Page`.

(WebCore::OpportunisticTaskScheduler::makeDeferralScope):

Helper method to make a `OpportunisticTaskDeferralScope`, which increments the deferral count (and
only decrements the count upon detroying the scope).

(WebCore::OpportunisticTaskScheduler::runLoopObserverFired):

Call into page to `performOpportunisticallyScheduledTasks`, once the runloop observer fires (i.e.,
the runloop has become idle) and only if we still have remaining time until the next deadline (which
is established when we were most recently rescheduled). This ensures that if post-rendering-update
tasks take a long time to finish and another rendering update is imminent, we&apos;ll skip this
opportunistic task call entirely, and instead do the work when we&apos;re less busy.

(WebCore::OpportunisticTaskScheduler::incrementDeferralCount):
(WebCore::OpportunisticTaskScheduler::decrementDeferralCount):

Private methods to manage the task deferral count; while the count is non-zero, we avoid calling
into the page to perform opportunistic tasks. When the deferral count is restored to 0, we then
wait for the runloop to become idle (via `RunLoopObserver`), and then run tasks if appropriate.

* Source/WebCore/page/OpportunisticTaskScheduler.h: Added.
(WebCore::OpportunisticTaskScheduler::create):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didCommitLoad):
(WebCore::Page::didFirstMeaningfulPaint):

Hold on to a `OpportunisticTaskDeferralScope` before the page has finished first paint.

(WebCore::Page::renderingUpdateCompleted):

Reschedule opportunistic tasks (see above for more information).

(WebCore::Page::performOpportunisticallyScheduledTasks):

Leave an empty method stub for now; in the next patch, this will call into a new `JSC::VM`
entrypoint to perform some pre-scheduled, periodic cleanup tasks, such as running the incremental
sweeper until the deadline.

* Source/WebCore/page/Page.h:
(WebCore::Page::opportunisticTaskScheduler const):
* Source/WebCore/platform/RunLoopObserver.h:
(WebCore::RunLoopObserver::RunLoopObserver):
(WebCore::RunLoopObserver::isRepeating const):
* Source/WebCore/platform/cf/RunLoopObserverCF.cpp:
(WebCore::RunLoopObserver::schedule):

Add an option to make the `RunLoopObserver` one-shot (as opposed to repeating). We use this flag
when creating runloop observers in the new `OpportunisticTaskScheduler`, since we don&apos;t want the
runloop observer to continue listening forever.

Canonical link: <a href="https://commits.webkit.org/265178@main">https://commits.webkit.org/265178@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3cd2a90f22bc2b761e2ab2478bb74fc9dc76ddc9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10117 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10361 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10614 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11768 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10126 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12351 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10312 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12717 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10272 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11037 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8550 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12151 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8341 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9166 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16476 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9443 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9318 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12577 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9799 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8958 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2426 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13196 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9603 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/OpportunisticTaskScheduler.cpp:
```diff
@@ -0,0 +1,110 @@
+        m_runLoopObserver->schedule();
+}
+
+std::unique_ptr<OpportunisticTaskDeferralScope> OpportunisticTaskScheduler::defer()
```
I'd call this makeDeferralScope instead for clarity.

rniwa commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2015,6 +2024,9 @@ void Page::renderingUpdateCompleted()
     }
+
+    if (!isUtilityPage())
+        m_opportunisticTaskScheduler->reschedule(m_lastRenderingUpdateTimestamp + Seconds(1. / m_displayNominalFramesPerSecond.value_or(60)));
```
We always assume 60fps by default?
Based on FIXME in WebPageProxy::swapToProvisionalPage, macOS will always use this default value?

whsieh replied:
Oh, that's a good point — I think that 60 is effectively the default value, since this is plumbed down from the UI process, which otherwise falls back to `FullSpeedFramesPerSecond` (60 fps) in the case where the display link is null for any reason:

```
FramesPerSecond DisplayLink::nominalFramesPerSecondFromDisplayLink(CVDisplayLinkRef displayLink)
{
    CVTime refreshPeriod = CVDisplayLinkGetNominalOutputVideoRefreshPeriod(displayLink);
    if (!refreshPeriod.timeValue)
        return FullSpeedFramesPerSecond;

    FramesPerSecond result = round((double)refreshPeriod.timeScale / (double)refreshPeriod.timeValue);
    return result ?: FullSpeedFramesPerSecond;
}
```

In practice, I don't think there's really a valid scenario where a (non-utility) `Page` would have a null `m_displayNominalFramesPerSecond` — we can probably just gate the rescheduling on having a non-null `m_displayNominalFramesPerSecond`:

```
void Page::renderingUpdateCompleted()
{
…
    if (!isUtilityPage() && m_displayNominalFramesPerSecond)
        m_opportunisticTaskScheduler->reschedule(m_lastRenderingUpdateTimestamp + Seconds(1. / *m_displayNominalFramesPerSecond));
}
```

rniwa commented on Source/WebCore/page/DOMTimer.cpp:
```diff
@@ -416,6 +425,30 @@ std::optional<MonotonicTime> DOMTimer::alignedFireTime(MonotonicTime fireTime) c
     return adjustedFireTime - (adjustedFireTime % alignmentInterval) + alignmentInterval + randomizedOffset;
 }
 
+void DOMTimer::takeOpportunisticTaskDeferralScopeIfPossible(ScriptExecutionContext& context)
```
This creates a new deferral scope, not "take" away an existing scope so it's probably better called makeOpportunisticTaskDeferralScopeIfPossible or createOpportunisticTaskDeferralScopeIfPossible.

whsieh replied:
Makes sense — I'll rename this to `makeOpportunisticTaskDeferralScopeIfPossible`

rniwa commented on Source/WebCore/page/DOMTimer.cpp:
```diff
@@ -416,6 +425,30 @@ std::optional<MonotonicTime> DOMTimer::alignedFireTime(MonotonicTime fireTime) c
+    m_opportunisticTaskDeferralScope = page->opportunisticTaskScheduler().defer();
+}
+
+void DOMTimer::resetOpportunisticTaskDeferralScope()
```
clearOpportunisticTaskDeferralScopeIfPossible?

rniwa commented on Source/WebCore/page/OpportunisticTaskScheduler.h:
```diff
@@ -0,0 +1,76 @@
+    void incrementDeferralCount();
+    void decrementDeferralCount();
+
+    Page& m_page;
```
Can we use WeakPtr instead?

whsieh replied:
Changed to `WeakPtr<Page>`

rniwa commented on Source/WebCore/page/OpportunisticTaskScheduler.cpp:
```diff
@@ -0,0 +1,110 @@
+OpportunisticTaskScheduler::OpportunisticTaskScheduler(Page& page)
+    : m_page(page)
+    , m_runLoopObserver(makeUnique<RunLoopObserver>(RunLoopObserver::WellKnownOrder::PostRenderingUpdate, [weakThis = WeakPtr { this }] {
+        if (weakThis)
```
We should store this into a Ref/RefPtr before calling runLoopObserverFired.

whsieh replied:
Good catch, fixed!

whsieh committed:
Add a mechanism to schedule opportunistic work in between rendering updates
https://bugs.webkit.org/show_bug.cgi?id=258032
rdar://110076652

Reviewed by Ryosuke Niwa.

Add an `OpportunisticTaskScheduler` helper class to be used in a subsequent patch to dispatch
periodically-scheduled cleanup tasks, such as incremental sweeping, in between rendering updates
(i.e. after the runloop begins to wait for more work, but before the next rendering update is
scheduled to start).

This `OpportunisticTaskScheduler` is essentially a wrapper around a `RunLoopObserver`, which calls
into `Page` to perform scheduled work until a given deadline. The `Page` is the only entity that
owns the task scheduler, and will tell it to reschedule after every rendering update. Additionally,
we introduce another mechanism — `OpportunisticTaskDeferralScope` — that prevents the task scheduler
from firing during certain time intervals where we know that scheduled work is imminent. For now,
this just includes:

1. Scheduled one-shot timers with a very short duration.
2. Any time prior to first paint, after navigation.

No change in behavior (yet).

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/ScriptExecutionContext.h:
(WebCore::ScriptExecutionContext::addTimeout):
(WebCore::ScriptExecutionContext::takeTimeout):
(WebCore::ScriptExecutionContext::removeTimeout): Deleted.

Change `removeTimeout` to `takeTimeout`, and have it return the removed `DOMTimer`. This allows us
to (more easily) call `clearOpportunisticTaskDeferralScopeIfPossible()` on the removed timer below.

* Source/WebCore/page/DOMTimer.cpp:
(WebCore::DOMTimer::install):
(WebCore::DOMTimer::removeById):
(WebCore::DOMTimer::fired):
(WebCore::DOMTimer::didStop):
(WebCore::DOMTimer::makeOpportunisticTaskDeferralScopeIfPossible):
(WebCore::DOMTimer::clearOpportunisticTaskDeferralScopeIfPossible):

Add a opportunistic task deferral scope to prevent us from scheduling opportunistic tasks in the
case where one-shot zero-delay timers are about to fire.

* Source/WebCore/page/DOMTimer.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::fireLayoutRelatedMilestonesIfNeeded):

Add some plumbing to notify `Page` when we hit first meaningful paint.

* Source/WebCore/page/OpportunisticTaskScheduler.cpp: Added.
(WebCore::OpportunisticTaskDeferralScope::OpportunisticTaskDeferralScope):
(WebCore::OpportunisticTaskDeferralScope::~OpportunisticTaskDeferralScope):

Add a RAII helper class to increment and decrement the deferral count in
`OpportunisticTaskScheduler`, which prevents opportunistically scheduled tasks from firing; this
RAII pattern hopefully makes it a bit easier to ensure that nothing ends up with unbalanced calls to
increment or decrement.

(WebCore::OpportunisticTaskScheduler::OpportunisticTaskScheduler):
(WebCore::OpportunisticTaskScheduler::reschedule):

Helper method to reschedule the task scheduler, with a target deadline; called after completing
rendering updates in `Page`.

(WebCore::OpportunisticTaskScheduler::makeDeferralScope):

Helper method to make a `OpportunisticTaskDeferralScope`, which increments the deferral count (and
only decrements the count upon detroying the scope).

(WebCore::OpportunisticTaskScheduler::runLoopObserverFired):

Call into page to `performOpportunisticallyScheduledTasks`, once the runloop observer fires (i.e.,
the runloop has become idle) and only if we still have remaining time until the next deadline (which
is established when we were most recently rescheduled). This ensures that if post-rendering-update
tasks take a long time to finish and another rendering update is imminent, we'll skip this
opportunistic task call entirely, and instead do the work when we're less busy.

(WebCore::OpportunisticTaskScheduler::incrementDeferralCount):
(WebCore::OpportunisticTaskScheduler::decrementDeferralCount):

Private methods to manage the task deferral count; while the count is non-zero, we avoid calling
into the page to perform opportunistic tasks. When the deferral count is restored to 0, we then
wait for the runloop to become idle (via `RunLoopObserver`), and then run tasks if appropriate.

* Source/WebCore/page/OpportunisticTaskScheduler.h: Added.
(WebCore::OpportunisticTaskScheduler::create):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::didCommitLoad):
(WebCore::Page::didFirstMeaningfulPaint):

Hold on to a `OpportunisticTaskDeferralScope` before the page has finished first paint.

(WebCore::Page::renderingUpdateCompleted):

Reschedule opportunistic tasks (see above for more information).

(WebCore::Page::performOpportunisticallyScheduledTasks):

Leave an empty method stub for now; in the next patch, this will call into a new `JSC::VM`
entrypoint to perform some pre-scheduled, periodic cleanup tasks, such as running the incremental
sweeper until the deadline.

* Source/WebCore/page/Page.h:
(WebCore::Page::opportunisticTaskScheduler const):
* Source/WebCore/platform/RunLoopObserver.h:
(WebCore::RunLoopObserver::RunLoopObserver):
(WebCore::RunLoopObserver::isRepeating const):
* Source/WebCore/platform/cf/RunLoopObserverCF.cpp:
(WebCore::RunLoopObserver::schedule):

Add an option to make the `RunLoopObserver` one-shot (as opposed to repeating). We use this flag
when creating runloop observers in the new `OpportunisticTaskScheduler`, since we don't want the
runloop observer to continue listening forever.

Canonical link: https://commits.webkit.org/265178@main

smfr commented on Source/WebCore/page/DOMTimer.h:
```diff
@@ -95,6 +99,7 @@ class DOMTimer final : public RefCounted<DOMTimer>, public SuspendableTimerBase,
     bool m_oneShot;
     Seconds m_currentTimerInterval;
     RefPtr<UserGestureToken> m_userGestureTokenToForward;
+    std::unique_ptr<OpportunisticTaskDeferralScope> m_opportunisticTaskDeferralScope;
```
Shame to bloat `DOMTimer` with this. Could we store this out of band?

smfr commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -5891,6 +5891,11 @@ void LocalFrameView::fireLayoutRelatedMilestonesIfNeeded()
+
+            if (m_frame->isMainFrame()) {
+                if (auto* page = m_frame->page())
+                    page->didFirstMeaningfulPaint();
```
Not quite right, as discussed on Slack.

whsieh replied:
Indeed — this should go into `firePaintRelatedMilestonesIfNeeded()`, when we actually finish the meaningful paint.

smfr commented on Source/WebCore/page/Page.cpp:
```diff
@@ -2013,6 +2022,9 @@ void Page::renderingUpdateCompleted()
         m_unfulfilledRequestedSteps = { };
     }
+
+    if (!isUtilityPage() && m_displayNominalFramesPerSecond)
```
`m_displayNominalFramesPerSecond` can be 120fps on a ProMotion display but we'll only run rendering updates at `preferredRenderingUpdateFramesPerSecond()`, which is probably what you want to use here?

whsieh replied:
Good point!

(I also noticed `preferredRenderingUpdateInterval` — would that be more appropriate here, or should I stick to `preferredRenderingUpdateFramesPerSecond`?)

smfr replied:
Those are equivalent (one is FPS, one is frame duration)

smfr commented on Source/WebCore/page/OpportunisticTaskScheduler.cpp:
```diff
@@ -0,0 +1,111 @@
+    if (!deadline)
+        return;
+
+    auto currentTime = MonotonicTime::now();
```
I wonder if using `ApproximateTime` would be good enough here, and cheaper.

whsieh replied:
Yeah, I think `ApproximateTime` should be okay given the timescales we're generally working with (on the order of milliseconds). Let me try that..
